File: doc-examples/r2dbc-example-java/src/main/java/example/FlightRepository.java
package example;

import io.micronaut.data.annotation.Upsert;
import io.micronaut.data.model.query.builder.sql.Dialect;
import io.micronaut.data.r2dbc.annotation.R2dbcRepository;
import io.micronaut.data.repository.CrudRepository;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

// tag::upsert-repository[]
@R2dbcRepository(dialect = Dialect.POSTGRES)
public interface FlightRepository extends CrudRepository<Flight, String> {

    @Upsert
    Mono<Flight> upsertMono(Flight flight);

    @Upsert
    Flux<Flight> upsertFlux(Iterable<Flight> flights);
}
// end::upsert-repository[]

File: doc-examples/r2dbc-example-python/src/main/python/example/FlightRepository.py
from micronaut.data.annotation import Upsert
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.r2dbc.annotation import R2dbcRepository
from micronaut.data.repository import CrudRepository
from reactor.core.publisher import Flux, Mono

from example.Flight import Flight


# tag::upsert-repository[]
@R2dbcRepository(dialect=Dialect.POSTGRES)
class FlightRepository(CrudRepository[Flight, str]):

    @Upsert
    def upsertMono(self, flight: Flight) -> Mono[Flight]: ...

    @Upsert
    def upsertFlux(self, flights: list[Flight]) -> Flux[Flight]: ...
# end::upsert-repository[]
